src/analyses/basePriv.ml

```diff
@@ -943,16 +979,22 @@ struct
     (* TODO: what about G_m globals in cpa that weren't actually written? *)
     CPA.fold (fun x v (st: BaseComponents (D).t) ->
         if is_protected_by ask m x then ( (* is_in_Gm *)
-          (* Extra precision in implementation to pass tests:
-             If global is read-protected by multiple locks,
-             then inner unlock shouldn't yet publish. *)
-          if not Param.check_read_unprotected || is_unprotected_without ask ~write:false x m then
-            sideg (V.protected x) v;
-          if atomic then
-            sideg (V.unprotected x) v; (* Publish delayed unprotected write as if it were protected by the atomic section. *)
-
+          (* Only apply sides for values that were actually written to globals!
+             This excludes invariants inferred through guards. *)
+          begin match D.precise_side x v st.priv with
```

Comment:
This seems like a separate improvement. I'm a bit unsure about it.

I first thought this should already be covered, but it's not because the implementation doesn't have the $$W$$ set that's mentioned in the traces paper. Instead we've abused $$\mathrm{dom}(\texttt{CPA})$$ for it, but that indeed doesn't cut it. An `invariant` write is supposed to add the variable to $$P$$, but not $$W$$, but the map domain based implementation makes it impossible.

Wouldn't it be simpler to just add the $$W$$ set into the implementation?
In that case I also don't see much reason to keep the current approach. I think protection is the only privatization which abuses the domain of `CPA` for some other purpose.

Reply:
The difference likely is that this only side-effects the actually written values whereas #1693 side-effects the joined value, so this here is somehow a cheap way to get what kind of amounts to path-sensitivity in `W`.

I think it may be worth paying this additional price, it was critical in the experimental evaluation here iirc.

Reply:
I am a bit hesitant about having this tangentially-related change in the same PR, but this extra abstraction is clean enough that it's OK to have around.
I currently don't have the time to more deeply think about this and #1693.